### PR TITLE
javascript-javascript1-week2

### PR DESCRIPTION
**Pull request by maribelpreite**
Hi mentors! Sorry for the late submission., it's been a crazy working week. 

I'm not sure if the candy exercise is accurate, especially when it says that we have to use the boughtCandies array to indicate if the user can buy more candies or not. As you'll see, I created a variable called totalAmount and then compared it to the amountToSpend variable. Even though the code works, I can't think of any other way to do it using the array.

Thank you all for your help and asistance! <3

### CODE REVIEW

File: javascript/javascript1/week2/weather_wear.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+function whatToWearBasedOnTemp (temperature) {
```
Comment (jason-vasilev) [inline javascript/javascript1/week2/weather_wear.js:1]: This exercise is great. Just remember to format your files now and then with Prettier. You can also change your VS Code settings so it runs Prettier each time you save a file - https://www.digitalocean.com/community/tutorials/how-to-format-code-with-prettier-in-visual-studio-code

File: javascript/javascript1/week2/weather_wear.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    } else if (temperature<18){
+        return "some fleece";
+    } else if (temperature<26){
+        return "some short sleeves";
+    } else {
+        return "some shorts";
+    }
+}
+
+const clothesToWear = whatToWearBasedOnTemp(18);
```
Comment (jason-vasilev) [inline javascript/javascript1/week2/weather_wear.js]: This works well. A step better could be creating a variable that is generated through `Math.random` set in a way to return a number between -20 and + 50 for example.

File: javascript/javascript1/week2/student_manager.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
+const class07Students = ["Anna", "Jonas", "Micah", "Noah", "Louise"];
+function addStudentToClass(studentName) {
+  // You write code here
+    if (class07Students.length>5 && studentName !== "Queen Margrethe") {
+        console.log("Cannot add more students to class 07");
+    } else if (class07Students.includes(studentName)) {
+        console.log("Student " + studentName + " is already in the class");
```
Comment (jason-vasilev) [inline javascript/javascript1/week2/student_manager.js]: You can use [template literals/strings](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) for cleaner look. 
```suggestion
        console.log(`Student ${studentName} is already in the class`);
```

File: javascript/javascript1/week2/student_manager.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
+const class07Students = ["Anna", "Jonas", "Micah", "Noah", "Louise"];
+function addStudentToClass(studentName) {
+  // You write code here
```
Comment (jason-vasilev) [inline javascript/javascript1/week2/student_manager.js]: You can remove this comment - it was just to indicate where you have to start.

File: javascript/javascript1/week2/student_manager.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    if (class07Students.length>5 && studentName !== "Queen Margrethe") {
+        console.log("Cannot add more students to class 07");
+    } else if (class07Students.includes(studentName)) {
```
Comment (jason-vasilev) [inline javascript/javascript1/week2/student_manager.js:11]: Well done, using the right method 👌

File: javascript/javascript1/week2/student_manager.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    } else if (studentName === "") {
+        console.log("Please add the student's name"); 
+    } else {
+        class07Students.push(studentName);
+    }
+}
+
+function getNumberOfStudents(class07Students) {
+  // You write code here
+    console.log("There are " + class07Students.length + " in class 07. See the full list of students below:");
```
Comment (jason-vasilev) [inline javascript/javascript1/week2/student_manager.js]: Would be nice if you have moved line 22 `console.log(class07Students);` inside here, instead of breaking it outside of the function.

File: javascript/javascript1/week2/student_manager.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
+const class07Students = ["Anna", "Jonas", "Micah", "Noah", "Louise"];
```
Comment (jason-vasilev) [inline javascript/javascript1/week2/student_manager.js]: Add 1-2 more students here to trigger the `console.log` from the check for reaching the limit of students in a class.

File: javascript/javascript1/week2/flight_booking.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
+function getFullName (firstname, surname, useFormalName, gender) {
```
Comment (jason-vasilev) [inline javascript/javascript1/week2/flight_booking.js]: The assignment has written the variables in wrong case (should be cammelCase). Could you fix them in your code?
`firstname` has to be `firstName`
`fullname1` has to be `fullName1`
etc.

File: javascript/javascript1/week2/flight_booking.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    if (useFormalName === true && gender === "female") {
```
Comment (jason-vasilev) [inline javascript/javascript1/week2/flight_booking.js]: Each time you check if a variable is true, you can write it shorter as this:
```
if (useFormalName && gender === "female") {
```

File: javascript/javascript1/week2/flight_booking.js [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+        return "Lady " + firstname + " " + surname;
```
Comment (jason-vasilev) [inline javascript/javascript1/week2/flight_booking.js]: You've repeated `firstname + " " + surname` 3 times. It might be a good idea to store it in a variable and use it instead, so you can avoid repetition.

File: javascript/javascript1/week2/flight_booking.js [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+            return "Lord " + firstname + " " + surname;
+        } else if (firstname !== "" && surname !== "") {
+        return firstname + " " + surname;
+        } else if (firstname === "" && surname !== "") {
+            return "Passenger " + surname;
+        } else if (firstname !== "" && surname === "") {
+            return "Passenger " + firstname;
+        } else if (firstname === "" && surname === "") {
+            return "Unknown Passenger";
+        }
```
Comment (jason-vasilev) [inline javascript/javascript1/week2/flight_booking.js]: I'd like you to rethink this logic with the following considerations:
- It shouldn't be allowed for a passenger to buy a plain ticket by writing just their first or last name. Both names should be provided. If either of them is missing, there should be an error message logged.
- Make sure empty characters from the name are removed. For example with the current logic it's possible to pass `firstName = " "` and `surname = " "`. The method you should use to check this is called [`trim()` ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/trim)

File: javascript/javascript1/week2/event_application.js [hunk 12]
```diff
@@ -0,0 +1,16 @@
+const daysSinceToday = [
```
Comment (jason-vasilev) [inline javascript/javascript1/week2/event_application.js:1]: It's preferred to start the week with "Monday" so the results would be more predictable

File: javascript/javascript1/week2/event_application.js [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+    "Friday",
+    "Saturday",
+    "Sunday",
+    "Monday",
+    "Tuesday",
+    "Wednesday"
+]
+
+function getEventWeekday(number) {
+    return daysSinceToday[number%7];
```
Comment (jason-vasilev) [inline javascript/javascript1/week2/event_application.js]: Unfortunately, this code will not give you the right results. In the task it's required to dynamically get today's day (index), by using the `[new Date()` object](https://www.w3schools.com/jsref/jsref_date_new.asp).
Then you need to add to today's day index (lets say Saturday would be 6 the days till the event (`number` in your case). The result should be passed with the modulo of 7.

```
function getEventWeekday(number) {
  const todayIs = new Date().getDay();
  return daysSinceToday[(todayIs + number) % 7];
}
```

File: javascript/javascript1/week2/candy_helper.js [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+    } else if (candyType === "chewing-gum") {
+        boughtCandies.push(weight*0.03);
+    }
+}
+
+function totalPrice () {
+    for (let i=0; i<boughtCandies.length; i++) {
+        totalAmount += boughtCandies[i];
+    }
+    console.log("Final Price: $" + totalAmount);
```
Comment (jason-vasilev) [inline javascript/javascript1/week2/candy_helper.js]: This is exactly how it had to be done. There are definitely other ways, but in this case this is the most common approach. Well done.

**Review by jason-vasilev — APPROVED** [review]
Hi, Maribel.

You actually have done the candy calculation task very well. Else, I've left small comments here and there. You've done well by making multiple commits, just remember to use Prettier, too.
See you tomorrow.

File: javascript/javascript1/week2/candy_helper.js (hunk 14 above)
Comment (maribelpreite) [reply javascript/javascript1/week2/candy_helper.js]: Hi, Yasen. First of all, thank you for taking the time to make all those comments. I truly appreciate your help. I've read your suggestions through and I've applied all of them. 

One of the things I'm not really sure about yet is how to make sure Prettier works, as I don't see much difference. I've taken a look at the link you sent and set Prettier to format on save. 

Regarding the event_application exercise, I've changed the array starting from Sunday, not Monday, as I noticed new Date().getDay() returns 0 for Sunday. I think if I had chosen Monday to be the first element of the array, the code wouldn't have worked properly. Please, correct me if I'm wrong. 

When it comes to the student_manager exercise, I've tried to add more students WITHIN the class07students array, and even though I won't be able to add more students by calling the function and passing new arguments, the array does accept more than 6 students outside of the function. This is absolutely expectable, but I'm not sure if it follows the task's instructions. 
EDIT: To solve this I created a new function using the .splice( ) method to delete all the elements from the array starting from the 7th.

A question I've come across while doing these exercises is the when to choose console.log( ) or return in a function. Is it correct to say that if I don't want the output of that function to be shown, I should use return?

I'll come to you today to check all of this, so don't worry about answering :) 
Thank you again for all the positive comments on the code! 💪💪💪

Comment (jason-vasilev) [reply javascript/javascript1/week2/candy_helper.js]: Sorry, I saw your message late. We could have went through it in class, but here is my reply.

A quick way to test if Prettier works is to remove all indentation from a file and then save it. If the file suddenly adds back the correct indentation, then Prettier works. Or if you add extra empty lines after the last line of code - Prettier would remove all, but the last one empty.

That's a very good reason and argument for the event_application to keep Sunday as the first day of the week. All good.

We use a lot of console.log now, but the idea is only for us as developers to test that the code works and we cover all edge cases. When you work on a real project, you remove all console logs at the end, as you don't want to give away what or how your code works exactly. Functions are often called through a variable, which then is console-logged. In this case, it doesn't make a big difference if you are console logining to the function or outside of it. Rasmus, however, suggested that we avoid as much as possible logging in functions, as you never know when they will be called and how many times, as that might clutter the terminal. Sorry, that I can't give a more definite answer.

I didn't see a problem with the student task. I tried to add a couple more students with the function and it worked fine. The task instructions were to add students in a way that you can test the various edge cases. I see you test for the limit of students "Cannot add more students to class 07" and "There are 7 in class 07. See the full list of students below: ...". May be add few more instances like those

```
addStudentToClass("Micah");
addStudentToClass("Steve");
addStudentToClass("Jason");
addStudentToClass("    ");
addStudentToClass("");
addStudentToClass("Bob");
addStudentToClass("Jason");
```
so you can log all other possible cases.

Comment (maribelpreite) [reply javascript/javascript1/week2/candy_helper.js]: Thanks for your reply Yasen! My biggest question was regarding return and console.log in a function, and after last Sunday it's much clearer! 

See you tomorrow! 🙌🏻
